resources/2025-11-11-secrets-redactdata.md

```diff
@@ -0,0 +1,756 @@
+
+
+### Non goals
+
+1. **Prevent all temporary storage**: The design accepts that encrypted sensitive data will temporarily exist for a brief period (milliseconds to seconds) between frontend save and backend decryption/redaction, before recipe execution begins.
+
+
+### User scenarios (optional)
+
+#### User story 1: Developer deploying secrets to Azure Key Vault
+
+Alice, a platform engineer, wants to deploy application secrets to Azure Key Vault using Radius. She defines a `Radius.Security/secrets` resource with sensitive credentials (property marked with `x-radius-sensitive` annotation) and associates it with an Azure Key Vault recipe. Radius encrypts the sensitive data. When the async worker processes the request, it decrypts and redacts the sensitive data from the database before executing the recipe. The recipe deploys the secrets to Azure Key Vault using the in-memory decrypted data. Alice can reference the secrets via output resources, but sensitive values are not persisted in Radius database.
+
+#### User story 2: Multi-environment secret management
+
+Bob manages multiple environments (dev, staging, prod) and needs different secret backends for each. He uses the same `Radius.Security/secrets` resource definition but associates different recipes: Kubernetes secrets for dev, HashiCorp Vault for staging, and Azure Key Vault for prod. The redaction mechanism works consistently across all backends without requiring backend-specific code.
+
+## User Experience (if applicable)
+
+### Resource Type Schema Definition
+
+The infrastructure operator defines a RRT schema with a `data` property annotated with `x-radius-sensitive` to indicate it contains sensitive information:
+
+```yaml
+# Excerpt of RRT schema using `x-radius-sensitive` annotation for sensitive fields
+types:
+  userTypeSensitive:    
+    apiVersions:
+      "2023-10-01-preview":
+        schema:
+          type: "object"
+          properties:
+            environment:
+              type: string
+              description: Environment ID for the userTypeSensitive resource
+            kind:
+              type: string
+              enum: [generic]
+              description: Type of data being stored
+            data:
+              type: string
+              x-radius-sensitive: true  # ← Marks this field as sensitive
+              description: Sensitive data to be deployed
+            required:
+              - environment
+              - data 
+```                     
+### Resource in Bicep: 
+
+Application developers use the resource type in their deployments:
+```bicep
+
+@secure
+param apiKey string
+
+resource appSensitive 'Radius.Security/userTypeSensitive@2023-10-01-preview' = {
+  name: 'app-sensitive'
+  properties: {
+    environment: environment.id
+    kind: 'generic'
+    data: apiKey
+  }
+}
+
+/*
+Note: x-radius-sensitive maps to appropriate secure Bicep type:
+   - `type: string` + `x-radius-sensitive: true` → `secureString`
```

Comment:
What happens if a user tries to add the tag onto an data type that's not a string or object?